### ISSUE TITLE
Review python versions for 1.14

### ISSUE THREAD

**Issue by klmcwhirter** [sympy/sympy#27341]
The WIP [Release Notes for 1.14](https://github.com/sympy/sympy/wiki/Release-Notes-for-1.14) shows the same supported Python versions as 1.13 - Python 3.8, 3.9, 3.10, 3.11, 3.12 and PyPy.

However, the [Python version support policy](https://github.com/sympy/sympy/wiki/Python-version-support-policy) suggest it be reviewed.

Please note that per the [python.org Versions](https://devguide.python.org/versions/#versions) page:
* 3.8 was EOL as of 2024-10-07
* keep 3.12 as the preferred version for now; unless there is a compelling reason to move to 3.13
* 3.13 is in wide use already; having been officially released on 2024-10-07
* 3.14 will be released in Oct 2025

*Is there a proposed release schedule for SymPy 1.14 yet?*

**Please note** that I do not use a python distro like conda. So I am not in a great position to analyze dependencies with those. Feedback from someone familiar with at least conda is welcome.

To start the conversation, I would like to propose that we:

* at least drop testing for 3.8
* add 3.13 to the matrix of tested versions (note I have been using 3.13 for a couple months with sympy and have not encountered any issues)
* think about 3.14 only if the release of SymPy 1.14 is in or after Oct 2025

I could make myself available to help with this.

I am aware that there are other conversations in flight that may impact this decision - e.g., #25155.

**Comment by oscarbenjamin** [comment]
There was a discussion about SPEC 0 here:
https://groups.google.com/g/sympy/c/VNz2xJ1Sywo/m/AWWqTOILBAAJ?utm_medium=email&utm_source=footer

If we follow that then I think we drop 3.8, 3.9 and 3.10 right now.

> Is there a proposed release schedule for SymPy 1.14 yet?

No, not yet.

> * add 3.13 to the matrix of tested versions

It is tested but under the `-dev` marker because it was added to track prereleases:
https://github.com/sympy/sympy/blob/fbe647fb34f90de01946fcaa1314207f11e5b26e/.github/workflows/runtests.yml#L314
I guess we just need to change it to `3.13` and `3.14-dev`.

**Comment by klmcwhirter** [comment]
That was the point of this issue exactly - review the complexity around Python, sympy's other dependencies against Python versions (e.g., tensorflow).

Just a reminder issue. Looks like some work (>1 story point).

Yahoo Mail: Search, Organize, Conquer 
 
  On Tue, Dec 3, 2024 at 3:56 AM, Oscar ***@***.***> wrote:   


There was a discussion about SPEC 0 here:
https://groups.google.com/g/sympy/c/VNz2xJ1Sywo/m/AWWqTOILBAAJ?utm_medium=email&utm_source=footer

If we follow that then I think we drop 3.8, 3.9 and 3.10 right now.


Is there a proposed release schedule for SymPy 1.14 yet?


No, not yet.

   
   - add 3.13 to the matrix of tested versions


It is tested but under the -dev marker because it was added to track prereleases:
https://github.com/sympy/sympy/blob/fbe647fb34f90de01946fcaa1314207f11e5b26e/.github/workflows/runtests.yml#L314
I guess we just need to change it to 3.13 and 3.14-dev.

**Comment by amirebrahimi** [comment]
> think about 3.14 only if the release of SymPy 1.14 is in or after Oct 2025

If the planned release of SymPy 1.14 is in October, would it be possible to get a minor release before then with contributed fixes?We've been in a holding pattern with releasing our own internal package since it depends on fixes post 1.13.3. The cadence used to be roughly 1 month per release.

**Comment by klmcwhirter** [comment]
@oscarbenjamin 

That was the point of this issue exactly - review the complexity around Python, sympy's other dependencies against Python versions (e.g., tensorflow).

I don't remember where I ran into them, but different items (including optional features) have a hard dependency on certain Python versions. I think it was a mixture of trying to run the tests and building that I kept running into things I needed to work around. Eventually I just bit the bullet and installed 3.12. I only had 3.13 installed at the time.

*3.13.1 is installed by default with Fedora WS 41.*

Just a reminder issue. Looks like some work (>1 story point).

*Wow, what a mess my phone email client makes of these replies. Sorry about that.*


  On Tue, Dec 3, 2024 at 3:56 AM, Oscar ***@***.***> wrote:   
There was a discussion about SPEC 0 here:
https://groups.google.com/g/sympy/c/VNz2xJ1Sywo/m/AWWqTOILBAAJ?utm_medium=email&utm_source=footer

If we follow that then I think we drop 3.8, 3.9 and 3.10 right now.

Is there a proposed release schedule for SymPy 1.14 yet?


No, not yet.

   
   - add 3.13 to the matrix of tested versions
It is tested but under the -dev marker because it was added to track prereleases:
https://github.com/sympy/sympy/blob/fbe647fb34f90de01946fcaa1314207f11e5b26e/.github/workflows/runtests.yml#L314
I guess we just need to change it to 3.13 and 3.14-dev.

**Comment by oscarbenjamin** [comment]
> If the planned release of SymPy 1.14 is in October, would it be possible to get a minor release before then with contributed fixes?

There is no planned release in October. I would like to get a release out ASAP but I am too busy right now to do it. Note that making a minor release is also more work than it sounds. Chances are CI is already broken on the release branch which has been inactive since September. Particular fixes would need to be cherry-picked etc. Release notes pages on the wiki have to be updated and so on.

The release process is more involved than it ideally would be largely because of the way that branches and release notes are handled (see gh-26169). Once the release branch is created it is relatively easy to push a tag to trigger the releases although each one involves fiddling around with the release notes in the wiki as well.

Working through the release blockers is a lot of work (see https://github.com/sympy/sympy/milestone/66). Once the release candidates are out then there will be bug reports and fixing those is a lot of work. If you look at the release history (https://pypi.org/project/sympy/#history) then you can see that it took 4 prereleases over a period of one month to get to 1.13 final. Then a bugfix 1.13.1 release was out a week or so later. Then more bugfix releases over some time.

Basically all of that is a lot of work, the majority of which is often done by me. Starting Monday I am going to be very busy teaching for 12 weeks straight and I don't even want to initiate the release process at all if the expectation is that I am personally going to do anything close to the amount of work that I did for 1.13.

If it required less commitment from me then that would be fine. Pretty much all of this can be done by other people e.g.:

- Go through all the release blockers and fix or defer them.
- Create all the necessary release notes pages.
- Update the AUTHORS file.
- Prepare the PRs that are needed when creating the release branch.
- Fix up all the versions in the CI scripts so that CI works both on master and on the release branch.
- Add the version constraint for mpmath and possibly update the version constraints for python-flint.

Something we have not previously done that would make it easier for making future maintenance releases:

- Pin versions of requirements in CI e.g. for building the docs on the release branch and also for linters and other things.

(It would be easier for the release process if CI requirements were generally pinned and updated by dependabot.)

Once we have a release branch it is relatively easy to push out a prerelease but that will generate new bug reports so next step (which also does not need to be done by me) is to address all of those bug reports probably with PRs to master and backport PRs cherry-picking for the release branch. Same process applies after final release.

If other people did all of those things then all that would be required from me would be a ~2hr process each time a new pre/final release is needed in order to push the tag, fiddle around with the release notes in the wiki and send an announcement email to the mailing list. That is something that I can commit to but if the expectation is that I am personally going to fix the blockers, respond to any of the bug reports etc then I probably don't even want to think about this for at least 12 weeks.

**Comment by oscarbenjamin** [comment]
The administrative part of the release process would be a lot easier if we switched from using the release notes bot to using something like towncrier. Designing how that is supposed to work is also a lot of work especially since it changes the basic PR workflow and requires some broader discussion. This is also something that anyone else could do (i.e. it doesn't have to be me). 

The release process would be much smoother if the release notes and the tools for managing them were all in the codebase rather than being in the release notes bot and on the wiki.

**Comment by oscarbenjamin** [comment]
I would also be fine with making a maintenance release with some particular fixes backported to the 1.13 branch. Again though others can do most of the necessary bits. Look at the wiki and the 1.13 branch history and 1.13 milestone PRs:

- Make an announcement on the mailing list with a request for what fixes people would like to see backported.
- Create release notes pages on the wiki.
- Bump the necessary versions on the branch.
- Backport whatever is needed to fix CI on the branch.
- Open backport PRs with whatever is going to be included in the release.

**Comment by oscarbenjamin** [comment]
See gh-25155 for discussion about changing the release notes system. My simple idea is just to require that each PR adds a file called `changelog/12345.txt`. Others want to make the PR workflow nicer than that but making it any nicer dramatically increases the complexity of the required tooling.

**Comment by klmcwhirter** [comment]
> See [gh-25155](https://github.com/sympy/sympy/issues/25155) for discussion about changing the release notes system. My simple idea is just to require that each PR adds a file called `changelog/12345.txt`. Others want to make the PR workflow nicer than that but making it any nicer dramatically increases the complexity of the required tooling.

Agreed. I spent a good portion of my 40+ year career on what is now known as CI/CD. We just called it build automation back when. Oh, the RCS and Perforce days ... shiver.

Simpler is better! When an over complicated pipeline breaks it tends to make people really angry. Because, Murphy's Law, it always breaks at the worst possible time. Ask me how I know !

It may help to remind them of the flow ...

1. Contributor sees something that can be improved and creates an issue
2. Creates a fork
3. Works on the contribution in a feature branch
4. Pushes up the feature branch
5. Then creates a PR ...
6. etc.

That feature branch could already contain the artifacts you mention and they can be validated just like the code, and formatted as needed for inclusion into the appropriate files. Once that is done those artifacts *could* be removed from the source tree.

That would be better than the template approach in use today IMHO. KISS - I love it.

Or the publishing of the release notes could be deferred to the time of release. The down side of doing that is the risk of someone inadvertently changing someone else's changelog, release notes, etc.

I have also automated release note generation by using the commit log. But that approach carries a pretty hefty proofing process by the build manager. Not ideal. It is better that happens during code review (PR) when the details are fresh on everyone's mind.

Keep It Stupidly Simple is my guiding principle ...

**Comment by amirebrahimi** [comment]
> See [gh-25155](https://github.com/sympy/sympy/issues/25155) for discussion about changing the release notes system. My simple idea is just to require that each PR adds a file called `changelog/12345.txt`. Others want to make the PR workflow nicer than that but making it any nicer dramatically increases the complexity of the required tooling.

Thanks for the link. I scanned through the discussion and couldn't find any mention of using a single `CHANGELOG.md` according to [Keep a Changelog](https://keepachangelog.com/en/1.1.0/). Has this been considered? Yes, it's manual, but then we have a task in our PRs that has to be checked off: "Did you update the changelog?" This has worked well for us across many projects. We use `sphinx` for docs and use an `mdinclude` in a corresponding `CHANGELOG.rst`.

**Comment by oscarbenjamin** [comment]
> using a single `CHANGELOG.md`

I think that there would be constant merge conflicts with everyone trying to edit the same part of the same file. That is why I suggested `changelog/12345.txt`: each PR adds the changelog in a unique file (e.g. the PR number). At release time or periodically a script can collect all the changelog entries into a single file which can be done in a semi-automatic way e.g. there can be a new PR where everyone can review the full collections of release notes before release.

**Comment by klmcwhirter** [comment]
> > using a single `CHANGELOG.md`
> 
> I think that there would be constant merge conflicts with everyone trying to edit the same part of the same file. That is why I suggested `changelog/12345.txt`: each PR adds the changelog in a unique file (e.g. the PR number). At release time or periodically a script can collect all the changelog entries into a single file which can be done in a semi-automatic way e.g. there can be a new PR where everyone can review the full collections of release notes before release.

AND the review of the changelog/release notes is greatly simplified. They are reviewed in the same tool (GitHub PR) with the code. That is key!

**Comment by amirebrahimi** [comment]
> I think that there would be constant merge conflicts with everyone trying to edit the same part of the same file.

With the different sections of the changelog (e.g. "Added", "Changed", "Fixed") we don't encounter as many conflicts as one might expect, but when there are conflicts they are easy to resolve. I'm just thinking that with all the frustrations you mentioned that a low-tech solution might alleviate some of those.

**Comment by oscarbenjamin** [comment]
> when there are conflicts they are easy to resolve.

Many SymPy contributors are not experienced with git so resolving any conflicts should not be assumed easy.

This is actually how we have the current release notes system: the current system is as easy as possible for contributors who are inexperienced with git even though it is probably not intuitive to anyone who is experienced with git and might not expect that editing the OP in a PR makes the release notes happen.

**Comment by amirebrahimi** [comment]
> Many SymPy contributors are not experienced with git so resolving any conflicts should not be assumed easy.

Good to know. I hadn't considered that.

**Comment by amirebrahimi** [comment]
> I would also be fine with making a maintenance release with some particular fixes backported to the 1.13 branch. Again though others can do most of the necessary bits. Look at the wiki and the 1.13 branch history and 1.13 milestone PRs:
> 
> * Make an announcement on the mailing list with a request for what fixes people would like to see backported.
> * Create release notes pages on the wiki.
> * Bump the necessary versions on the branch.
> * Backport whatever is needed to fix CI on the branch.
> * Open backport PRs with whatever is going to be included in the release.

I'm going to be working on releasing a bugfix version 1.13.4 this sprint and will follow these steps, so any assistance during the next two weeks would be much appreciated.

**Comment by oscarbenjamin** [comment]
See https://groups.google.com/g/sympy/c/myCZrHVveZs/m/sxsRRnGgCgAJ

Closing in favour of gh-27939

**Comment by oscarbenjamin** [comment]
We test numpy on Python 3.9 in CI so it should be fine:
https://github.com/sympy/sympy/blob/089edf0ea752497120cbaa1a5af4b5d4bcab9a85/.github/workflows/runtests.yml#L146-L192
Often fixes are needed in sympy because of changes in numpy but generally the fix still works with older numpy versions. There is a CI job that tests numpy nightly wheels so we basically track numpy's main branch as well as current release for each supported Python version.

**Comment by klmcwhirter** [comment]
I just wanted to be sure you were aware. I ran into this in another FOSS project.
